### PR TITLE
Allow swapping out the memory allocation functions.

### PR DESCRIPTION
**Pull request by arthurschreiber**
This is mainly useful for bindings like rugged, where we can swap these
functions with the ones provided by Ruby. This helps Ruby to better
track memory usage through libgit2, and allows it to trigger GC runs
more intelligently.

The implementation in this pull request uses the existing `git_libgit2_opts` infrastructure to allow changing the allocator functions.

This is obviously missing documentation, as I want to first get some feedback whether `git_libgit2_opts` is the correct place for this or not. 😅

I tested this from the rugged bindings, and things were working just fine, without any super obvious performance impact. 👍

### CODE REVIEW

File: src/util.h [hunk 1]
```diff
@@ -82,28 +82,34 @@
 
 #else
 
+void* (*git__malloc_func)(size_t size) = malloc;
+void* (*git__calloc_func)(size_t nitems, size_t size) = calloc;
```
Comment (arthurschreiber) [inline src/util.h]: I think we'll want to have `typedef`s for these, so we don't need to repeat them here and over in `src/settings.c`.

**Comment by ethomson** [comment]
Don’t we have a `strndup` also?  I think that we probably want to either wire that up for consistency or drop this `strdup` and always implement our `strdup` in terms of `malloc` and the user need not reimplement it.

File: src/util.h [hunk 2]
```diff
@@ -82,28 +82,34 @@
...
+char* (*git__strdup_func)(const char *str1) = strdup;
+void* (*git__realloc_func)(void *ptr, size_t size) = realloc;
+void* (*git__free_func)(void *ptr) = free;
```
Comment (carlosmn) [inline src/util.h]: This returns void, not void pointer

File: src/util.h [hunk 3]
```diff
@@ -82,28 +82,34 @@
...
+
```
Comment (carlosmn) [inline src/util.h]: All of these assignments need to be in a .c file. The header can then have an extern definition

**Comment by pks-t** [comment]
Haha. Should've looked first if there was another PR already. Well, anyway. see #4576 for another implementation.

**Comment by carlosmn** [comment]
I do like the approach of passing in a structure; it should be less error prone than a bunch of untyped varargs.

**Comment by carlosmn** [comment]
#4576 also brings up that we have things like `git__reallocarray` which are implemented on top of the standard allocator with better semantics. I think we should let the user specify those but we should make them optional so we fall back to our own `reallocarray` if the pointer is `NULL`. This would mean another pass through them to make sure that e.g. `git__stdalloc_reallocarray` doesn't call `realloc` but `git__stdalloc_realloc` to use the names from the other PR.

**Comment by pks-t** [comment]
I'm closing this in favor of #4576. Thanks for your work?

**Pull request closed without merging**